### ISSUE TITLE
Request preemphasis and deemphasis modules

### ISSUE THREAD

**Issue by AppleHolic** [pytorch/audio#226]
## 🚀 Feature
I request Preemphasis / Deemphasis modules. [wiki](https://en.wikipedia.org/wiki/Preemphasis_improvement) In several sound related deep-learning tasks, These module are often used to help sound quality.

## Motivation

In my speech enhancement case, that model usually generate high frequency noises without preemphasis. With this method, it brings many helps to enhance enhancement quality. But, to use of it on training models, it should be implemented with pytorch.

## Pitch and Alternatives

I already did code first version of preemphasis and deemphasis written in pytorch. 
So, after checking out below code has some bugs, I suggest to add this modules to transforms.py

- [Permalink of my github repo](https://github.com/AppleHolic/pytorch_sound/blob/b741b994795830c35c9627117d76727c6c4bf550/pytorch_sound/models/sound.py#L64)

- code snippet

```python
class PreEmphasis(torch.nn.Module):

    def __init__(self, coef: float = 0.97):
        super().__init__()
        self.coef = coef
        # make kernel
        # In pytorch, the convolution operation uses cross-correlation. So, filter is flipped.
        self.register_buffer(
            'flipped_filter', torch.FloatTensor([-self.coef, 1.]).unsqueeze(0).unsqueeze(0)
        )

    def forward(self, input: torch.tensor) -> torch.tensor:
        assert len(input.size()) == 3, 'The number of dimensions of input tensor must be 3!'
        # reflect padding to match lengths of in/out
        input = F.pad(input, (1, 0), 'reflect')
        return F.conv1d(input, self.flipped_filter)


class InversePreEmphasis(torch.nn.Module):
    """
    Implement Inverse Pre-emphasis by using RNN to boost up inference speed.
    """

    def __init__(self, coef: float = 0.97):
        super().__init__()
        self.coef = coef
        self.rnn = torch.nn.RNN(1, 1, 1, bias=False, batch_first=True)
        # use originally on that time
        self.rnn.weight_ih_l0.data.fill_(1)
        # multiply coefficient on previous output
        self.rnn.weight_hh_l0.data.fill_(self.coef)

    def forward(self, input: torch.tensor) -> torch.tensor:
        x, _ = self.rnn(input.transpose(1, 2))
        return x.transpose(1, 2)
```

### Sample Plots

- scipy codes

```python
from scipy import signal

def preemphasis(x, coeff=0.97):
    return signal.lfilter([1, -coeff], [1], x).astype(np.float32)


def inv_preemphasis(x, coeff=0.97):
    return signal.lfilter([1], [1, -coeff], x).astype(np.float32)
```

- Sample wave

![image](https://user-images.githubusercontent.com/3727290/62818610-7be20d00-bb84-11e9-8d72-2c3d7730018d.png)

- Preemphasis / deemphasis sample

![image](https://user-images.githubusercontent.com/3727290/62818626-b51a7d00-bb84-11e9-882e-6975c12c2693.png)

![image](https://user-images.githubusercontent.com/3727290/62818629-c19ed580-bb84-11e9-8cad-567cf88eb8cb.png)

- Pytorch samples

![image](https://user-images.githubusercontent.com/3727290/62818721-f5c6c600-bb85-11e9-889c-215f64deb748.png)

**Comment by vincentqb** [comment]
This is relevant for #260 :)

**Comment by yyd19948** [comment]
> ## 🚀 Feature
> I request Preemphasis / Deemphasis modules. [wiki](https://en.wikipedia.org/wiki/Preemphasis_improvement) In several sound related deep-learning tasks, These module are often used to help sound quality.
> 
> ## Motivation
> In my speech enhancement case, that model usually generate high frequency noises without preemphasis. With this method, it brings many helps to enhance enhancement quality. But, to use of it on training models, it should be implemented with pytorch.
> 
> ## Pitch and Alternatives
> I already did code first version of preemphasis and deemphasis written in pytorch. So, after checking out below code has some bugs, I suggest to add this modules to transforms.py
> 
> * [Permalink of my github repo](https://github.com/AppleHolic/pytorch_sound/blob/b741b994795830c35c9627117d76727c6c4bf550/pytorch_sound/models/sound.py#L64)
> * code snippet
> 
> ```python
> class PreEmphasis(torch.nn.Module):
> 
>     def __init__(self, coef: float = 0.97):
>         super().__init__()
>         self.coef = coef
>         # make kernel
>         # In pytorch, the convolution operation uses cross-correlation. So, filter is flipped.
>         self.register_buffer(
>             'flipped_filter', torch.FloatTensor([-self.coef, 1.]).unsqueeze(0).unsqueeze(0)
>         )
> 
>     def forward(self, input: torch.tensor) -> torch.tensor:
>         assert len(input.size()) == 3, 'The number of dimensions of input tensor must be 3!'
>         # reflect padding to match lengths of in/out
>         input = F.pad(input, (1, 0), 'reflect')
>         return F.conv1d(input, self.flipped_filter)
> 
> 
> class InversePreEmphasis(torch.nn.Module):
>     """
>     Implement Inverse Pre-emphasis by using RNN to boost up inference speed.
>     """
> 
>     def __init__(self, coef: float = 0.97):
>         super().__init__()
>         self.coef = coef
>         self.rnn = torch.nn.RNN(1, 1, 1, bias=False, batch_first=True)
>         # use originally on that time
>         self.rnn.weight_ih_l0.data.fill_(1)
>         # multiply coefficient on previous output
>         self.rnn.weight_hh_l0.data.fill_(self.coef)
> 
>     def forward(self, input: torch.tensor) -> torch.tensor:
>         x, _ = self.rnn(input.transpose(1, 2))
>         return x.transpose(1, 2)
> ```
> 
> ## Additional context
> ### Sample Plots
> * scipy codes
> 
> ```python
> from scipy import signal
> 
> def preemphasis(x, coeff=0.97):
>     return signal.lfilter([1, -coeff], [1], x).astype(np.float32)
> 
> 
> def inv_preemphasis(x, coeff=0.97):
>     return signal.lfilter([1], [1, -coeff], x).astype(np.float32)
> ```
> 
> * Sample wave
> 
> ![image](https://user-images.githubusercontent.com/3727290/62818610-7be20d00-bb84-11e9-8d72-2c3d7730018d.png)
> 
> * Preemphasis / deemphasis sample
> 
> ![image](https://user-images.githubusercontent.com/3727290/62818626-b51a7d00-bb84-11e9-882e-6975c12c2693.png)
> 
> ![image](https://user-images.githubusercontent.com/3727290/62818629-c19ed580-bb84-11e9-8cad-567cf88eb8cb.png)
> 
> * Pytorch samples
> 
> ![image](https://user-images.githubusercontent.com/3727290/62818721-f5c6c600-bb85-11e9-889c-215f64deb748.png)

I repeat the preempsis and deempsis，but it can not recover the orignal signal.

**Comment by carolineechen** [comment]
addressed in #2871
